macOS 11: fix for remake (Coq platform 8.12 package versions)

MSoegtropIMC opened the pull request:
This PR fixes an issue with remake on macOS 11 / XCode 12.
It also fixes a missing dependency for coq-gappa.

There are also newer versions of some packages affected which will be fixed in a separate PR for Coq platform 8.13 package versions (for testability reasons).

@silene : FYI

gares commented:
You need to add bison and flex to
https://github.com/coq/opam-coq-archive/blob/2d33beb975de4ef2cd2761f6a0703b26a60cde18/.gitlab-ci.yml#L20

MSoegtropIMC commented:
> You need to add bison and flex to

I don't think so. gappa depends on flex and bison, but not coq-gappa - it depends on gappa.

See: (https://github.com/ocaml/opam-repository/blob/master/packages/gappa/gappa.1.3.5/opam)

MSoegtropIMC commented:
Ah you mean to the .yml file - not the opam file.

Doesn't this use depext to install prerequisites? If not, why not?

gares commented:
Hem, I guess the answer is that it is perfectible ;-)

gares commented:
(also, I'm not so sure depext was there from the very start)

MSoegtropIMC commented:
OK, let me see if I can change it to use depext.

silene commented on released/packages/coq-gappa/coq-gappa.1.4.4/opam:
```diff
@@ -17,6 +17,7 @@ depends: [
   "ocaml"
   "coq" {>= "8.8.1" & < "8.13~"}
   "coq-flocq" {>= "3.0"}
+  "gappa" {>= "1.3.5"}
```
Please do not add this dependency. The presence of `gappa` is not needed to build `coq-gappa`, and updating `gappa` should not cause a recompilation of `coq-gappa`. (So, it is not even a `depopts`.) If you need to express the relationship between versions, use instead
```
conflicts: [
  "gappa" {< "1.3" }
]
```
but that would not bring much (by lack of such packages).

MSoegtropIMC replied:
I thought about this. The reason why I added the dependency is that the plugin is not usable without gappa. So while it is not a build dependency, it is a runt time dependency. I think that users can reasonably expect that coq-gappa does install the runtime dependencies it needs. But I agree that it is a disadvantage for CI, but only if CI does not test if the resulting plugin actually works - which I think CI should do.

silene replied:
We could certainly change `build` to
```
build: [
  ["autoconf"] {dev}
  ["./configure"]
  ["./remake" "-j%{jobs}%"]
  ["./remake" "check"] {with-test}
]
```
and then it would make sense to have
```
depends: [
  "gappa" {with-test}
]
```

But that is an orthogonal matter. (And does the CI actually call `opam install --with-test`?)

erikmd replied:
good point: it might be useful to run tests as well in opam-coq-archive.
FWIW, if tweaking the CLI command itself is cumbersome, it suffices to add the environment variable
```
OPAMWITHTEST: "true"
```
(we actually noticed this with @Zimmi48 when we recently added support for [environment variables in docker-coq-action](https://github.com/coq-community/docker-coq-action#export))

MSoegtropIMC replied:
@silene : My point is that the main purpose of the opam packages is installation by users and not convenience in test systems. I don't see a situation where a user would like to install coq-gappa and not gappa. And a user is also not likely to explicitly request tests. If we want to make it such that in tests coq-gappa compiles without gappa for CI performance reasons, then it should be done such that it is explicitly removed in CI testing. Not sure how one would do that - I guess we would need some quick-test flag.

Zimmi48 replied:
> updating `gappa` should not cause a recompilation of `coq-gappa`.

This is a legitimate concern though. It seems that opam supports this use case with `post` dependencies (run-time only dependencies IIUC): see https://opam.ocaml.org/doc/Manual.html#opamfield-depends.

silene replied:
Unfortunately, that is not how `post` works (which may explain why almost no package uses it, except OCaml itself, for which it was designed). When running Opam, there are two resolution stages. First, the set of needed actions (remove, install) is computed by calling a constraint solver. Second, the directed acyclic graph of needed actions is computed. The `post` variable is only used during this second stage, to decide whether to put some edges between vertices. So, as soon as you state that `gappa-coq` depends on `gappa` (whether `post` of not), the remove/install vertices for `gappa-coq` will be part of the graph when modifying `gappa`, since they have been found during the first resolution stage. The `post` part just means that there is no edge between the vertices for installing `gappa` and `gappa-coq`. So, their installation will happen in a random order.

MSoegtropIMC replied:
> updating gappa should not cause a recompilation of coq-gappa.

Compared to gappa, coq-gappa is quite light weight, so I don't see much of an issue with that. It is also not so that gappa changes every other week. The main issue I see is that people install coq-gappa and then it doesn't work.

So if the options are:
1.) people have to know that if they want to use gappa in Coq, they have to install gappa and coq-gappa
2.) coq-gappa is recompiled without need when gappa changes

I am for option 2. You might want to read (https://www.npr.org/2019/03/08/701417140/when-not-reading-the-fine-print-can-cost-your-soul) to get an idea on how many people read a "please install gappa if you install coq-gappa" note before they find that coq-gappa does not work - especially when coq-gappa is the prerequisite of some higher level automation.

palmskog commented:
@MSoegtropIMC do you still plan to update this pull request? I would really prefer to see a green CI run before merging. Just adding bison and flex to the yml file seems like the quickest option.

MSoegtropIMC commented:
@palmskog : I am discussing with @silene how to fix it. Removing the dependency obviously would also fix CI (since then it wouldn't need flex and bison at all). If we agree that the dependency should be kept, then I need to fix CI so that it works.

MSoegtropIMC commented:
@silene : can we come to a conclusion on this? My main argument is that for a user it does not make any sense to install coq-gappa without gappa, so coq-gappa should depend on it. Your argument (as far as I understand) is that this is inefficient for CI because coq-gappa can be CI tested independent of gappa (at least in some way).

My conclusion is that usability is more important than CI efficiency, especially for a stable package like gappa. In case you don't agree, please elaborate your arguments.

silene commented:
My point is that `gappa` is not needed at all to compile `coq-gappa`, and thus should not be marked as a build dependency of `coq-gappa`. So, as the maintainer of this package, I disapprove of this change.

But let us be blunt. I am not interested in making simpler the life of new users who would like to use the Coq support library for the Gappa tool despite having no clue what Gappa is. Most of my existing users have installed Gappa already (e.g., through their Linux distribution: https://pkgs.org/download/gappa), and I see no point in forcing them to compile it again. Also most of them have never even run an IDE for Coq and they do not care about the tactic. They only want to run `gappa -Bcoq | coqtop` and rightfully boast that their algorithm has the highest assurance level. In some cases, they do not even bother shipping the corresponding Gappa scripts (which might have been automatically generated anyway, e.g., by Metalibm). They directly ship the Coq certificates, in which case the recipient of these files only need `coq-gappa` but not `gappa` to verify them.

MSoegtropIMC commented:
I think this might change in the near future. E.g. it is not unlikely that @AndrewAppel gives examples for verifying floating point code with gappa in his lectures - he recently worked on such examples together with @ybertot. I personally find the plugin more user friendly and easy to use for applications which have their automation inside Coq and not outside of it. The solution Yves provided uses gappa explicitly afair, but I guess one would change this for a lecture to use the plugin.

Also I think one can argue that the dependencies given in an opam package are not build dependencies but dependencies needed to use or run something. Opam has the concept of a build only dependency (like gcc or bison) but not the concept of a run only dependency. So one can assume that something which is not explicitly a build dependency is a run dependency.

I also don't understand why saying that gappa is or should be used directly conflicts with this. I proposed to make coq-gappa dependent on gappa and not the other way around. If most of your users use just gappa, they won't see any difference at all.

But if you absolutely do not want this, it is not a big issue either. In the Coq Platform I can simply put coq-gappa and gappa in the package list. It is just inconvenient / confusing for people installing coq-gappa via opam.

Please consider my above arguments and let me know your final conclusion. I will then adjust the PR accordingly.

silene commented:
> Also I think one can argue that the dependencies given in an opam package are not build dependencies but dependencies needed to use or run something. Opam has the concept of a build only dependency (like gcc or bison) but not the concept of a run only dependency. So one can assume that something which is not explicitly a build dependency is a run dependency.

Your assumption is wrong; that is not how it works. Opam uses a two-stage process when considering dependencies. First, it filters the so-called "filtered package formulas" into "concrete package formulas". (The terminology is a bit weird, because "filtered" means that they have yet to be filtered.) All package formulas that are not false at this point are dependencies of the package. Opam then tries to satisfy them during a second stage.

So, if you write `"gappa" {>= "1.3.5"}`, this is a filtered package formula that gets filtered as itself. If you write, `"gappa" {>=  "1.3.5" & build}`, this a filtered package formula that gets filtered as `"gappa" {>= "1.3.5"}` if `coq-gappa` is being built, and as `"gappa" {false}` otherwise (in which case it is no longer a dependency, see above).

Thus, if you really wanted to express a runtime dependency, you would write `"gappa" {>= "1.3.5" & !built}`. Except that it does not work at all. Indeed, the first time you install `coq-gappa`, the package formula will evaluate to `"gappa" {false}` (since `coq-gappa` is being built), and thus the dependency is just ignored. So, the way Opam was designed makes it intrinsically impossible to have runtime dependencies, as far as I know.

> If most of your users use just gappa, they won't see any difference at all.

Unless they play some dirty tricks with their `PATH` environment variable, that will override their existing version of Gappa. That is a major difference!

> let me know your final conclusion

`coq-gappa` should not depend on `gappa`.

MSoegtropIMC commented:
> So, the way Opam was designed makes it intrinsically impossible to have runtime dependencies, as far as I know.

You explain that in opam there is no "runtime only" dependency - which would be called for here. I fully agree with this: there is a "build & run time" and a "build only" dependency. My argument is that from an opam user point of view it is better to have a superflous build time dependency than a missing run time dependency and for this reason if something is a "runtime only" dependency the best choice is to mark it as "build & run time" dependecy - which is what I did. But I am fine to delay further discussion on this until there are complaints that the coq-gappa package is broken. As I mentoned it is no issue for Coq Platform.

> coq-gappa should not depend on gappa.

I will change the PR.

MSoegtropIMC commented:
@palmskog @silene : FYI : I rebased and added the same remake patch to all newer versions of the affected packages (and I removed the coq-gappa -> gappa dependency).

palmskog commented:
@MSoegtropIMC there seems to be some kind of problem with applying the patches:
```
#=== ERROR while compiling coq-flocq.3.4.0 ====================================#
These patches didn't apply at /builds/coq/opam-coq-archive/opam-root-4.07.1-2.0.6-plain/4.07.1/.opam-switch/build/coq-flocq.3.4.0:
  - remake.patch: "/usr/bin/patch -p1 -i /builds/coq/opam-coq-archive/opam-root-4.07.1-2.0.6-plain/log/processed-patch-19844-1a1e04" exited with code 1

#=== ERROR while compiling coq-coquelicot.3.2.0 ===============================#
These patches didn't apply at /builds/coq/opam-coq-archive/opam-root-4.07.1-2.0.6-plain/4.07.1/.opam-switch/build/coq-coquelicot.3.2.0:
  - remake.patch: "/usr/bin/patch -p1 -i /builds/coq/opam-coq-archive/opam-root-4.07.1-2.0.6-plain/log/processed-patch-19844-f35e91" exited with code 1
```

MSoegtropIMC commented:
@palmskog : looks the patch is already in some of the new packages - I will check tomorrow.

MSoegtropIMC commented:
The solution to the mystery is that some packages already have one of the two patches, but not both. I adjusted the patches accordingly.

MSoegtropIMC commented:
@palmskog : this should be good to go now. It puts all packages using remake with versions used in Coq Platform 8.12 or newer to the same state (which works with recent XCode).
